### PR TITLE
Add OLMoE

### PR DESCRIPTION
**Pull request by Muennighoff**
# What does this PR do?

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

The model will be released in ~1 week - can we already review this so that we can merge right upon release?

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
Hey! Feel free to ping me for a review once ready!

**Comment by Muennighoff** [comment]
> Hey! Feel free to ping me for a review once ready!

It's ready! :) I will update the README & double-check the slow tests once the model is released if that's fine!

**Comment by Muennighoff** [comment]
@ArthurZucker would be great if we could get it reviewed soon 😇

**Comment by Muennighoff** [comment]
We'll release the model on Tuesday, would be amazing if we could have this approved by then!

**Comment by ArthurZucker** [comment]
Oups sorry reviewing now

File: src/transformers/models/olmoe/modeling_olmoe.py [hunk 1]
```diff
@@ -0,0 +1,1368 @@
...
+            base = self.base * (
+                (self.scaling_factor * seq_len / self.max_position_embeddings) - (self.scaling_factor - 1)
+            ) ** (self.dim / (self.dim - 2))
+            inv_freq = 1.0 / (
+                base ** (torch.arange(0, self.dim, 2, dtype=torch.int64).float().to(x.device) / self.dim)
+            )
+            self.register_buffer("inv_freq", inv_freq, persistent=False)  # TODO joao: this may break with compilation
+
+        cos, sin = super().forward(x, position_ids)
+        return cos, sin
```
Comment (ArthurZucker) [inline src/transformers/models/olmoe/modeling_olmoe.py:264]: the TODO should not be here, it's probably copied from Olmo?

File: src/transformers/models/olmoe/modeling_olmoe.py [hunk 2]
```diff
@@ -0,0 +1,1368 @@
...
+        `tuple(torch.Tensor)` comprising of the query and key tensors rotated using the Rotary Position Embedding.
+    """
+    cos = cos.unsqueeze(unsqueeze_dim)
+    sin = sin.unsqueeze(unsqueeze_dim)
+    q_embed = (q * cos) + (rotate_half(q) * sin)
+    k_embed = (k * cos) + (rotate_half(k) * sin)
+    return q_embed, k_embed
+
+
+class OlmoeMLP(nn.Module):
```
Comment (ArthurZucker) [inline src/transformers/models/olmoe/modeling_olmoe.py:303]: missing copied from (gemma2 for example_

File: src/transformers/models/olmoe/modeling_olmoe.py [hunk 3]
```diff
@@ -0,0 +1,1368 @@
...
+                f"hidden_size must be divisible by num_heads (got `hidden_size`: {self.hidden_size}"
+                f" and `num_heads`: {self.num_heads})."
+            )
+
+        self.q_proj = nn.Linear(self.hidden_size, self.num_heads * self.head_dim, bias=config.attention_bias)
+        self.k_proj = nn.Linear(self.hidden_size, self.num_key_value_heads * self.head_dim, bias=config.attention_bias)
+        self.v_proj = nn.Linear(self.hidden_size, self.num_key_value_heads * self.head_dim, bias=config.attention_bias)
+        self.o_proj = nn.Linear(self.hidden_size, self.hidden_size, bias=config.attention_bias)
+        self.q_norm = OlmoeRMSNorm(self.hidden_size)
+        self.k_norm = OlmoeRMSNorm((self.hidden_size // self.num_heads) * self.num_key_value_heads)
```
Comment (ArthurZucker) [inline src/transformers/models/olmoe/modeling_olmoe.py:366]: Cohere and Chameleon both have a similar layer norm

File: src/transformers/models/olmoe/modeling_olmoe.py [hunk 4]
```diff
@@ -0,0 +1,1368 @@
...
+                )
+            elif scaling_type == "dynamic":
+                self.rotary_emb = OlmoeDynamicNTKScalingRotaryEmbedding(
+                    self.head_dim,
+                    max_position_embeddings=self.max_position_embeddings,
+                    scaling_factor=scaling_factor,
+                    base=self.rope_theta,
+                )
+            else:
+                raise ValueError(f"Unknown RoPE scaling type {scaling_type}")
```
Comment (ArthurZucker) [inline src/transformers/models/olmoe/modeling_olmoe.py:394]: we no longer do this now! See llama for example

File: src/transformers/models/olmoe/modeling_olmoe.py [hunk 5]
```diff
@@ -0,0 +1,1368 @@
...
+
+    def forward(
```
Comment (ArthurZucker) [inline src/transformers/models/olmoe/modeling_olmoe.py:396]: is this copied from OLMO?

File: src/transformers/models/olmoe/modeling_olmoe.py [hunk 6]
```diff
@@ -0,0 +1,1368 @@
...
+
+        attn_output = self.o_proj(attn_output)
+
+        if not output_attentions:
+            attn_weights = None
+
+        return attn_output, attn_weights, past_key_value
+
+
+class OlmoeFlashAttention2(OlmoeAttention):
```
Comment (ArthurZucker) [inline src/transformers/models/olmoe/modeling_olmoe.py:462]: same here if copied from olmo

File: src/transformers/models/olmoe/modeling_olmoe.py [hunk 7]
```diff
@@ -0,0 +1,1368 @@
...
+
+
+OLMOE_ATTENTION_CLASSES = {
+    "eager": OlmoeAttention,
+    "flash_attention_2": OlmoeFlashAttention2,
+    "sdpa": OlmoeSdpaAttention,
+}
+
+
+class OlmoeSparseMoeBlock(nn.Module):
```
Comment (ArthurZucker) [inline src/transformers/models/olmoe/modeling_olmoe.py:670]: reviewing fast but probably missing copied from here as well no?

Comment (ArthurZucker) [reply src/transformers/models/olmoe/modeling_olmoe.py:670]: (Qwen2-moe)

File: src/transformers/models/olmoe/modeling_olmoe.py [hunk 8]
```diff
@@ -0,0 +1,1368 @@
...
+            current_hidden_states = expert_layer(current_state) * routing_weights[top_x, idx, None]
+
+            # However `index_add_` only support torch tensors for indexing so we'll use
+            # the `top_x` tensor here.
+            final_hidden_states.index_add_(0, top_x, current_hidden_states.to(hidden_states.dtype))
+        final_hidden_states = final_hidden_states.reshape(batch_size, sequence_length, hidden_dim)
+        return final_hidden_states, router_logits
+
+
+class OlmoeDecoderLayer(nn.Module):
```
Comment (ArthurZucker) [inline src/transformers/models/olmoe/modeling_olmoe.py:718]: sam ehere

File: src/transformers/models/olmoe/modeling_olmoe.py (hunk 7 above)
Comment (ArthurZucker) [reply src/transformers/models/olmoe/modeling_olmoe.py:670]: also I probably would not recommend you to use this formulation. As it will not be compatible with torch compile. #30793 gives explanation but if you have a conversion script, could be good to default with the correct format (I mean format that will allow ud to use the torch compile compatible formulation!)

File: src/transformers/models/olmoe/modeling_olmoe.py [hunk 9]
```diff
@@ -0,0 +1,1368 @@
...
+        return MoeCausalLMOutputWithPast(
+            loss=loss,
+            logits=logits,
+            past_key_values=outputs.past_key_values,
+            hidden_states=outputs.hidden_states,
+            attentions=outputs.attentions,
+            router_logits=outputs.router_logits,
+        )
+
+    def prepare_inputs_for_generation(
```
Comment (ArthurZucker) [inline src/transformers/models/olmoe/modeling_olmoe.py:1301]: missing copied from!

**Review by ArthurZucker — COMMENTED** [review]
Sorry for the late review!
Mostly missing copied from and should be good to go otherwise. 
Could go the extra mile to have compile compatible verseion of the MOE blcok!

File: src/transformers/models/olmoe/modeling_olmoe.py (hunk 7 above)
Comment (Muennighoff) [reply src/transformers/models/olmoe/modeling_olmoe.py:672]: Different from Qwen2-moe cuz we have no shared expert

File: src/transformers/models/olmoe/modeling_olmoe.py (hunk 8 above)
Comment (Muennighoff) [reply src/transformers/models/olmoe/modeling_olmoe.py:720]: Different from others cuz we have no shared expert

File: src/transformers/models/olmoe/modeling_olmoe.py (hunk 6 above)
Comment (Muennighoff) [reply src/transformers/models/olmoe/modeling_olmoe.py:461]: This is now copied from Llama but without
```
                "The attention layers in this model are transitioning from computing the RoPE embeddings internally "
                "through `position_ids` (2D tensor with the indexes of the tokens), to using externally computed "
                "`position_embeddings` (Tuple of tensors, containing cos and sin). In v4.45 `position_ids` will be "
                "removed and `position_embeddings` will be mandatory."
```
as I've directly changed the RoPE computation so not adding the Copy statement

File: src/transformers/models/olmoe/modeling_olmoe.py (hunk 4 above)
Comment (Muennighoff) [reply src/transformers/models/olmoe/modeling_olmoe.py]: Removed, ty!

File: src/transformers/models/olmoe/convert_olmoe_weights_to_hf.py [hunk 10]
```diff
@@ -0,0 +1,281 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
```
[the remark this answers is not in the document]
Comment (Muennighoff) [reply src/transformers/models/olmoe/convert_olmoe_weights_to_hf.py:1]: Is that mandatory? Maybe the entire header can just be removed? Seems redundant to have this header for every file when the license is clear from the repo..

File: src/transformers/models/olmoe/configuration_olmoe.py [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+            raise ValueError(
+                "`rope_scaling` must be a dictionary with two fields, `type` and `factor`, " f"got {self.rope_scaling}"
+            )
+        rope_scaling_type = self.rope_scaling.get("type", None)
+        rope_scaling_factor = self.rope_scaling.get("factor", None)
+        if rope_scaling_type is None or rope_scaling_type not in ["linear", "dynamic"]:
+            raise ValueError(
+                f"`rope_scaling`'s type field must be one of ['linear', 'dynamic'], got {rope_scaling_type}"
+            )
+        if rope_scaling_factor is None or not isinstance(rope_scaling_factor, float) or rope_scaling_factor <= 1.0:
```
[the remark this answers is not in the document]
Comment (Muennighoff) [reply src/transformers/models/olmoe/configuration_olmoe.py]: Changed!

File: src/transformers/models/olmoe/modeling_olmoe.py [hunk 12]
```diff
@@ -0,0 +1,1368 @@
...
+        # Compute the average probability of routing to these experts
+        router_prob_per_expert = torch.sum(routing_weights * router_per_expert_attention_mask, dim=0) / torch.sum(
+            router_per_expert_attention_mask, dim=0
+        )
+
+    overall_loss = torch.sum(tokens_per_expert * router_prob_per_expert.unsqueeze(0))
+    return overall_loss * num_experts
+
+
+class OlmoeRMSNorm(nn.Module):
```
[the remark this answers is not in the document]
Comment (Muennighoff) [reply src/transformers/models/olmoe/modeling_olmoe.py:184]: The default eps value is different (1e-5 instead of 1e-6)

File: src/transformers/models/olmoe/modeling_olmoe.py (hunk 1 above)
Comment (Muennighoff) [reply src/transformers/models/olmoe/modeling_olmoe.py]: Removed!

File: src/transformers/models/olmoe/modeling_olmoe.py [hunk 13]
```diff
@@ -0,0 +1,1368 @@
...
+            this tensor is not affected by padding. It is used to update the cache in the correct position and to infer
+            the complete sequence length.
+"""
+
+
+@add_start_docstrings(
+    "The bare Olmoe Model outputting raw hidden-states without any specific head on top.",
+    OLMOE_START_DOCSTRING,
+)
+class OlmoeModel(OlmoePreTrainedModel):
```
[the remark this answers is not in the document]
Comment (Muennighoff) [reply src/transformers/models/olmoe/modeling_olmoe.py:934]: Added!

File: src/transformers/models/olmoe/modeling_olmoe.py (hunk 3 above)
Comment (Muennighoff) [reply src/transformers/models/olmoe/modeling_olmoe.py:392]: Yeah but cannot use copy statement as I've changed to the new RoPE implementation and other things slightly differ

File: src/transformers/models/olmoe/modeling_olmoe.py (hunk 2 above)
Comment (Muennighoff) [reply src/transformers/models/olmoe/modeling_olmoe.py:329]: Added!

File: docs/source/en/model_doc/olmoe.md [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+⚠️ Note that this file is in Markdown but contain specific syntax for our doc-builder (similar to MDX) that may not be
+rendered properly in your Markdown viewer.
+
+-->
+
+# OLMoE
+
+## Overview
+
+The OLMoE model was proposed in [TODO](TODO) by TODO.
```
[the remark this answers is not in the document]
Comment (Muennighoff) [reply docs/source/en/model_doc/olmoe.md:21]: Thanks! Will do this as soon as released (very soon), maybe we can make sure all the rest is good first!

File: src/transformers/models/olmoe/modeling_olmoe.py (hunk 7 above)
Comment (Muennighoff) [reply src/transformers/models/olmoe/modeling_olmoe.py:677]: > also I probably would not recommend you to use this formulation. As it will not be compatible with torch compile. https://github.com/huggingface/transformers/pull/30793 gives explanation but if you have a conversion script, could be good to default with the correct format (I mean format that will allow ud to use the torch compile compatible formulation!)

There's no new & better formulation used by any model in transformers yet, right?

Comment (ArthurZucker) [reply src/transformers/models/olmoe/modeling_olmoe.py:677]: It is implemented in the linked PR here: https://github.com/huggingface/transformers/pull/30793/files#diff-733ab0a772c69f78b1d8ed361e6ae1fda7243652887aed0bab5d3ecf07794c01R789

File: src/transformers/models/olmoe/modeling_olmoe.py (hunk 6 above)
Comment (ArthurZucker) [reply src/transformers/models/olmoe/modeling_olmoe.py:466]: perfect thanks!

File: src/transformers/models/olmoe/modeling_olmoe.py (hunk 3 above)
Comment (ArthurZucker) [reply src/transformers/models/olmoe/modeling_olmoe.py]: yep no worries here!

File: src/transformers/models/olmoe/convert_olmoe_weights_to_hf.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/models/olmoe/convert_olmoe_weights_to_hf.py:1]: 😄 yeah it is redundant but the licence AFAIK but it's a nit don't worry

File: docs/source/en/model_doc/olmoe.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply docs/source/en/model_doc/olmoe.md:21]: okay!

File: src/transformers/models/olmoe/modeling_olmoe.py (hunk 7 above)
Comment (ArthurZucker) [reply src/transformers/models/olmoe/modeling_olmoe.py:677]: it's completely up to you!

Comment (Muennighoff) [reply src/transformers/models/olmoe/modeling_olmoe.py:677]: Oh thanks; I guess I was confused why the PR got closed - has it not been merged in anywhere?

Comment (Muennighoff) [reply src/transformers/models/olmoe/modeling_olmoe.py:677]: I took a look but it seems like there's a trade-off where the PR from @zhenglongjiepheonix increases memory usage for better speed while this implementation here requires less memory but is slower? Can you correct me if I'm wrong @zhenglongjiepheonix @ArthurZucker ?

Comment (zhenglongjiepheonix) [reply src/transformers/models/olmoe/modeling_olmoe.py:677]: It's mostly for compatibility of torch compile, the compiled model is 3x~4x faster in the case of Mixtral, have not made any comparisons on the eager mode

Comment (Muennighoff) [reply src/transformers/models/olmoe/modeling_olmoe.py:677]: That's great! But it requires more memory or not really?

Comment (zhenglongjiepheonix) [reply src/transformers/models/olmoe/modeling_olmoe.py:677]: Yes, it does

Comment (zhenglongjiepheonix) [reply src/transformers/models/olmoe/modeling_olmoe.py:677]: at least for training, I think the extra memory required is significant if we have large inputs

Comment (Muennighoff) [reply src/transformers/models/olmoe/modeling_olmoe.py:677]: Thanks for clarifying! Not worth doing it for this PR here then I think; For training people are best off using https://github.com/allenai/OLMo/pull/639

**Comment by ArthurZucker** [comment]
COuld you make sure you rebased and CIs are green? Will review again after that!

**Review by ArthurZucker — APPROVED** [review]
LGTM let's rebase and fix the CIs 🤗  (there is also a todo in the .md as well)

**Comment by Muennighoff** [comment]
Great, all passing! 🙌

**Comment by ArthurZucker** [comment]
Thanks for your contribution! 🔥

**Comment by kalomaze** [comment]
Is it a problem on my end if I notice that a higher native batch size during training results in higher losses? (With DeepSpeed)

image

Purple line is higher native bs FFT, orange line is gradient accumulation instead of a higher native batch size.
I'm guessing expert routing parallelism is maybe not properly handled for bs>1?

**Comment by ArthurZucker** [comment]
cc @Muennighoff would have a lot more clues than me 😉

**Comment by Muennighoff** [comment]
> Is it a problem on my end if I notice that a higher native batch size during training results in higher losses? (With DeepSpeed)
> 
> image
> 
> Purple line is higher native bs FFT, orange line is gradient accumulation instead of a higher native batch size.
> 
> I'm guessing expert routing parallelism is maybe not properly handled for bs>1?

Hm not sure about this - are other models the same for you in both scenarios? How about other MoEs?
